### ISSUE TITLE
Propagation SupportsJoin Issue

### ISSUE THREAD

**Issue by huangydyn** [openzipkin/brave#1058]
Original Issue refer to https://github.com/spring-cloud/spring-cloud-sleuth/issues/1523

# Describe
Spring Cloud version Greenwich.SR4

# Sample
use feign and sleuth, and config spring.sleuth.supports-join=false, parent spanid not correct during propagation.

## Invoker
Context
traceId spanId parentSpanId
eacdf1fa6276a833, eacdf1fa6276a833,null

## Receiver
Receive Http Header
x-b3-spanid, Value: 0ee7e9f326298f95
x-b3-parentspanid, Value: eacdf1fa6276a833
x-b3-traceid, Value: eacdf1fa6276a833

Context
traceId spanId parentSpanId
eacdf1fa6276a833, a3f53beaf4bcb48d, 0ee7e9f326298f95

The receiver parent spanid is not correct, it should be eacdf1fa6276a833 not 0ee7e9f326298f95, and x-b3-parentspanid has been ignored.

**Comment by adriancole** [comment]
there's no code related to this repo in this issue description. feign isn't here and supportsJoin is server side not client side (feign). can we become more concrete about a change that would affect code in this repo?

**Comment by huangydyn** [comment]
sample repo to replicate issue is in https://github.com/huangydyn/sleuth-demo.

demo trace chain is 'spring-server-enterpoint' -> 'spring-server-consumer-one' -> 'spring-server-consumer-two'.

All three services(client and server) have configured spring.sleuth.supports-join=false in application.yml.

After spring.sleuth.supports-join is configured to false, X-B3-ParentSpanId in MDC become not correct

**Comment by adriancole** [comment]
hi, the easiest way for us to investigate something is if you make a
fork of https://github.com/openzipkin/sleuth-webmvc-example and a
branch that changes the code to use the thing that you feel isn't
working. it sounds like it is adding feign and changing a property. if
you can do that, it is something I can take a look at. generally
ad-hoc examples we don't know work in the base case absorb time.

**Comment by huangydyn** [comment]
I've created forked repo https://github.com/huangydyn/sleuth-webmvc-example as to reproduce.

# What I changed
add spring.sleuth.supports-join=false in application.properties
change log format to: [%X{X-B3-TraceId:-},%X{X-B3-SpanId:-},%X{X-B3-ParentSpanId:-}] %-5p [%t] %C{2} - %m%n
# Reproduced Issue 
## Frontend
In Frontend, log is shown as:
```
[5e24016d67d450e261d9bfad46228d32,61d9bfad46228d32,] INFO  [http-nio-8081-exec-1] s.w.Frontend - FrondEnd receive request
```
SpanId is 61d9bfad46228d32
## Backend
In Backend, log is shown as:
```
[5e24016d67d450e261d9bfad46228d32,ff74aaca6544c404,903aa6f70b57af74] INFO  [http-nio-9000-exec-1] s.w.Backend -
 BackEnd receive request
```
Current ParentSpanId is not correct, it should be 61d9bfad46228d32 not 903aa6f70b57af74.

Thanks

**Comment by adriancole** [comment]
thanks.. looking into it

**Comment by adriancole** [comment]
TL;DR; everything is ok, it is just a matter of which log lines we are looking at.

I think what was confusing here was that the log statement `FrondEnd receive request` was in the server span, not the client one. A client is a child of the server span. If you set `logging.level.root=debug`, you can see the actual client span with id `659208af17cbf54b`. Note you can also see this if you were to look at the zipkin UI as opposed to the log statements.

```
2020-01-20 10:22:03.538 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] INFO  [http-nio-8081-exec-1] s.w.Frontend - FrondEnd receive request
 7930 --- [nio-8081-exec-1] sleuth.webmvc.Frontend                   : FrondEnd receive request
2020-01-20 10:22:03.550 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] DEBUG [http-nio-8081-exec-1] o.s.c.l.CompositeLog - HTTP GET http://localhost:9000/api
 7930 --- [nio-8081-exec-1] o.s.web.client.RestTemplate              : HTTP GET http://localhost:9000/api
2020-01-20 10:22:03.552 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] DEBUG [http-nio-8081-exec-1] o.s.c.l.CompositeLog - Accept=[text/plain, application/json, application/*+json, */*]
 7930 --- [nio-8081-exec-1] o.s.web.client.RestTemplate              : Accept=[text/plain, application/json, application/*+json, */*]
2020-01-20 10:22:03.561 [5e250ecbc6bef45f36e28326b2391765,659208af17cbf54b,36e28326b2391765] DEBUG [http-nio-8081-exec-1] s.u.l.i.LoggingProviderImpl$JULWrapper - sun.net.www.MessageHeader@45434b259 pairs: {GET /api HTTP/1.1: null}{Accept: text/plain, application/json, application/*+json, */*}{X-B3-TraceId: 5e250ecbc6bef45f36e28326b2391765}{X-B3-SpanId: 659208af17cbf54b}{X-B3-ParentSpanId: 36e28326b2391765}{X-B3-Sampled: 1}{User-Agent: Java/13.0.1}{Host: localhost:9000}{Connection: keep-alive}
```

Here, you can see that the header representing the client `X-B3-SpanId: 659208af17cbf54b` is in the log format for the parentID field as it should.
```
2020-01-20 10:22:03.651 [5e250ecbc6bef45f36e28326b2391765,54ccdc6c397d5194,659208af17cbf54b] INFO  [http-nio-9000-exec-1] s.w.Backend - BackEnd receive request header X-B3-TraceId : 5e250ecbc6bef45f36e28326b2391765,X-B3-SpanId: 659208af17cbf54b X-B3-ParentSpanId: 36e28326b2391765
 7932 --- [nio-9000-exec-1] sleuth.webmvc.Frontend                   : BackEnd receive request header X-B3-TraceId : 5e250ecbc6bef45f36e28326b2391765,X-B3-SpanId: 659208af17cbf54b X-B3-ParentSpanId: 36e28326b2391765
```

**Comment by huangydyn** [comment]
> TL;DR; everything is ok, it is just a matter of which log lines we are looking at.
> 
> I think what was confusing here was that the log statement `FrondEnd receive request` was in the server span, not the client one. A client is a child of the server span. If you set `logging.level.root=debug`, you can see the actual client span with id `659208af17cbf54b`. Note you can also see this if you were to look at the zipkin UI as opposed to the log statements.
> 
> ```
> 2020-01-20 10:22:03.538 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] INFO  [http-nio-8081-exec-1] s.w.Frontend - FrondEnd receive request
>  7930 --- [nio-8081-exec-1] sleuth.webmvc.Frontend                   : FrondEnd receive request
> 2020-01-20 10:22:03.550 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] DEBUG [http-nio-8081-exec-1] o.s.c.l.CompositeLog - HTTP GET http://localhost:9000/api
>  7930 --- [nio-8081-exec-1] o.s.web.client.RestTemplate              : HTTP GET http://localhost:9000/api
> 2020-01-20 10:22:03.552 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] DEBUG [http-nio-8081-exec-1] o.s.c.l.CompositeLog - Accept=[text/plain, application/json, application/*+json, */*]
>  7930 --- [nio-8081-exec-1] o.s.web.client.RestTemplate              : Accept=[text/plain, application/json, application/*+json, */*]
> 2020-01-20 10:22:03.561 [5e250ecbc6bef45f36e28326b2391765,659208af17cbf54b,36e28326b2391765] DEBUG [http-nio-8081-exec-1] s.u.l.i.LoggingProviderImpl$JULWrapper - sun.net.www.MessageHeader@45434b259 pairs: {GET /api HTTP/1.1: null}{Accept: text/plain, application/json, application/*+json, */*}{X-B3-TraceId: 5e250ecbc6bef45f36e28326b2391765}{X-B3-SpanId: 659208af17cbf54b}{X-B3-ParentSpanId: 36e28326b2391765}{X-B3-Sampled: 1}{User-Agent: Java/13.0.1}{Host: localhost:9000}{Connection: keep-alive}
> ```
> 
> Here, you can see that the header representing the client `X-B3-SpanId: 659208af17cbf54b` is in the log format for the parentID field as it should.
> 
> ```
> 2020-01-20 10:22:03.651 [5e250ecbc6bef45f36e28326b2391765,54ccdc6c397d5194,659208af17cbf54b] INFO  [http-nio-9000-exec-1] s.w.Backend - BackEnd receive request header X-B3-TraceId : 5e250ecbc6bef45f36e28326b2391765,X-B3-SpanId: 659208af17cbf54b X-B3-ParentSpanId: 36e28326b2391765
>  7932 --- [nio-9000-exec-1] sleuth.webmvc.Frontend                   : BackEnd receive request header X-B3-TraceId : 5e250ecbc6bef45f36e28326b2391765,X-B3-SpanId: 659208af17cbf54b X-B3-ParentSpanId: 36e28326b2391765
> ```



> TL;DR; everything is ok, it is just a matter of which log lines we are looking at.
> 
> I think what was confusing here was that the log statement `FrondEnd receive request` was in the server span, not the client one. A client is a child of the server span. If you set `logging.level.root=debug`, you can see the actual client span with id `659208af17cbf54b`. Note you can also see this if you were to look at the zipkin UI as opposed to the log statements.
> 
> ```
> 2020-01-20 10:22:03.538 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] INFO  [http-nio-8081-exec-1] s.w.Frontend - FrondEnd receive request
>  7930 --- [nio-8081-exec-1] sleuth.webmvc.Frontend                   : FrondEnd receive request
> 2020-01-20 10:22:03.550 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] DEBUG [http-nio-8081-exec-1] o.s.c.l.CompositeLog - HTTP GET http://localhost:9000/api
>  7930 --- [nio-8081-exec-1] o.s.web.client.RestTemplate              : HTTP GET http://localhost:9000/api
> 2020-01-20 10:22:03.552 [5e250ecbc6bef45f36e28326b2391765,36e28326b2391765,] DEBUG [http-nio-8081-exec-1] o.s.c.l.CompositeLog - Accept=[text/plain, application/json, application/*+json, */*]
>  7930 --- [nio-8081-exec-1] o.s.web.client.RestTemplate              : Accept=[text/plain, application/json, application/*+json, */*]
> 2020-01-20 10:22:03.561 [5e250ecbc6bef45f36e28326b2391765,659208af17cbf54b,36e28326b2391765] DEBUG [http-nio-8081-exec-1] s.u.l.i.LoggingProviderImpl$JULWrapper - sun.net.www.MessageHeader@45434b259 pairs: {GET /api HTTP/1.1: null}{Accept: text/plain, application/json, application/*+json, */*}{X-B3-TraceId: 5e250ecbc6bef45f36e28326b2391765}{X-B3-SpanId: 659208af17cbf54b}{X-B3-ParentSpanId: 36e28326b2391765}{X-B3-Sampled: 1}{User-Agent: Java/13.0.1}{Host: localhost:9000}{Connection: keep-alive}
> ```
> 
> Here, you can see that the header representing the client `X-B3-SpanId: 659208af17cbf54b` is in the log format for the parentID field as it should.
> 
> ```
> 2020-01-20 10:22:03.651 [5e250ecbc6bef45f36e28326b2391765,54ccdc6c397d5194,659208af17cbf54b] INFO  [http-nio-9000-exec-1] s.w.Backend - BackEnd receive request header X-B3-TraceId : 5e250ecbc6bef45f36e28326b2391765,X-B3-SpanId: 659208af17cbf54b X-B3-ParentSpanId: 36e28326b2391765
>  7932 --- [nio-9000-exec-1] sleuth.webmvc.Frontend                   : BackEnd receive request header X-B3-TraceId : 5e250ecbc6bef45f36e28326b2391765,X-B3-SpanId: 659208af17cbf54b X-B3-ParentSpanId: 36e28326b2391765
> ```

As I understand, `BackEnd receive request` is the child span of `FrondEnd receive request`
What confuse me is that 
- client( backend) received http header is `X-B3-SpanId: 659208af17cbf54b`, so  client span should be 659208af17cbf54b; And zipkin log shows spanId is 659208af17cbf54b which is correct I think

- But in client( backend) log MDC context, client span is shown as 54ccdc6c397d5194, and parentspan came to be 659208af17cbf54b.

So do you mean client log MDC context is not correct?

**Comment by adriancole** [comment]
when you disable join, you are saying do *not* use the same span ID on the server (backend) as the client ID. So, your last assumptions are the opposite which makes sense when you leave things default. since this is more chat in nature, let's follow-up on https://gitter.im/spring-cloud/spring-cloud-sleuth if you need more questions answered
